src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.HttpContentAsyncStream.cs

```diff
@@ -0,0 +1,412 @@
+                }
+
+                lock (_syncObj)
+                {
+                    // If we've been disposed, throw an exception so as to end the CopyToAsync operation.
+                    if (_disposed)
+                    {
+                        VerboseTrace("WriteAsync when already disposed");
+                        throw CreateHttpRequestException();
+                    }
+
+                    if (_buffer == null)
+                    {
+                        // There's no waiting reader.  Store everything for later.
+                        VerboseTrace("No waiting reader. Queueing.");
+                        Debug.Assert(_asyncOp == null, "No one else should be waiting");
+                        Update(false, buffer, offset, count, NewTcs());
+                        RegisterCancellation(cancellationToken);
+                        return _asyncOp.Task;
+                    }
+
+                    // There's a waiting reader.  We'll give them our data and store
+                    // for later anything they can't use.
+                    Debug.Assert(_waiterIsReader, "Concurrent WriteAsync calls are not permitted");
+                    Debug.Assert(_count > 0, "Expected reader to need at least 1 byte");
+                    Debug.Assert(_asyncOp != null, "Reader should have created a task to complete");
+
+                    // Complete the waiting reader with however much we can give them
+                    int bytesToCopy = Math.Min(count, _count);
+                    Buffer.BlockCopy(buffer, offset, _buffer, _offset, bytesToCopy);
+                    _cancellationRegistration.Dispose();
+                    bool completed = _asyncOp.TrySetResult(bytesToCopy);
+                    Debug.Assert(completed, "No one else should have completed the reader");
+                    VerboseTrace("Writer transferred " + bytesToCopy + " to reader.");
+
+                    if (bytesToCopy < count)
+                    {
+                        // If we have more bytes from this write, store the arguments
+                        // for use later and return a task to be completed by a reader.
+                        Update(false, buffer, offset + bytesToCopy, count - bytesToCopy, NewTcs());
+                        RegisterCancellation(cancellationToken);
+                        return _asyncOp.Task;
+                    }
+                    else
+                    {
+                        // We used up all of the data in this write, so complete
+                        // the caller synchronously.
+                        Update(false, null, 0, 0, null);
+                        return Task.CompletedTask;
+                    }
+                }
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                lock (_syncObj)
+                {
+                    // Mark ourselves as disposed, and make sure to wake
```

Comment:
Nit: I realize another call to Dispose won't hurt but it seems cleaner to check for _disposed==false before the cleanup